### PR TITLE
Adding clarifying documentation on the usage of second_order_closure

### PR DESCRIPTION
**Pull request by mortonjt**
Addresses https://github.com/PyTorchLightning/pytorch-lightning/issues/3539

The intention of this PR is to add documentation clarify the usage of the `second_order_closure` function.

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/optimizers.rst [hunk 1]
```diff
@@ -123,3 +123,33 @@ Here we add a learning-rate warm up
...
         optimizer.step()
         optimizer.zero_grad()
+
+
+----------
+
+Using the closure functions for optimization
+--------------------------------------------
+
+When using optimization schemes such as LBFGS, the `second_order_closure` needs to be enabled.  By default, this function is defined by wrapping the training_step and the backward steps as follows
```
Comment (Borda) [inline docs/source/optimizers.rst:133]:
```suggestion
When using optimization schemes such as LBFGS, the `second_order_closure` needs to be enabled.  By default, this function is defined by wrapping the `training_step` and the backward steps as follows
```

File: docs/source/optimizers.rst [hunk 2]
```diff
@@ -123,3 +123,33 @@ Here we add a learning-rate warm up
...
+        # on_after_backward callback
+        pl_module.on_after_backward(result.training_step_output, batch_idx, result.loss)
+
+        return result
+
+This default `second_order_closure` function can be enabled by passing it directly into the `optimizer.step`
+
+.. testcode::
+
+    def optimizer_step(self, current_epoch, batch_nb, optimizer, optimizer_idx, second_order_closure, on_tpu=False, using_native_amp=False, using_lbfgs=False):
```
Comment (Borda) [inline docs/source/optimizers.rst:153]:
```suggestion
    # This default `second_order_closure` function can be enabled by passing it directly into the `optimizer.step`
    def optimizer_step(self, current_epoch, batch_nb, optimizer, optimizer_idx, second_order_closure, on_tpu=False, using_native_amp=False, using_lbfgs=False):
```

**Review by Borda — APPROVED** [review]
LGTM :]

File: docs/source/optimizers.rst [hunk 3]
```diff
@@ -123,3 +123,30 @@ Here we add a learning-rate warm up
         # update params
         optimizer.step()
         optimizer.zero_grad()
+
+
```
Comment (rohitgr7) [inline docs/source/optimizers.rst:127]:
```suggestion
```

File: docs/source/optimizers.rst [hunk 4]
```diff
@@ -123,3 +123,30 @@ Here we add a learning-rate warm up
...
+----------
+
+Using the closure functions for optimization
+--------------------------------------------
+
+When using optimization schemes such as LBFGS, the `second_order_closure` needs to be enabled.  By default, this function is defined by wrapping the `training_step` and the backward steps as follows
```
Comment (rohitgr7) [inline docs/source/optimizers.rst:133]:
```suggestion
When using optimization schemes such as LBFGS, the `second_order_closure` needs to be enabled. By default, this function is defined by wrapping the `training_step` and the backward steps as follows
```

**Review by rohitgr7 — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]
LGTM

**Pull request merged**